File: JavaAndScala/src/main/scala/org/lang/scala/conclusion/design/factory/singleton/design/LazySingletonFactory.scala
package org.lang.scala.conclusion.design.factory.singleton.design

import scala.reflect._
import org.lang.scala.common.utils.ReflectionUtils

/**
 * 	This is a stand-alone object to implement lazy singleton factory pattern
 * 	-- Declare a private, volatile and Any-typed reference
 *  -- Declare a public and static getter to get this unique object:
 *     -- Double-checked locking pattern is used in this pattern
 *     
 *	@author VinceYuan
 */
object LazySingletonFactory {
  
  /*	Declare a private, volatile and Any-typed reference	 */
  @volatile private var instance: Any = _
  
  /**
   * 	This is a method to get the singleton instance for Scala only
   */
  def getInstance[T : ClassTag] = {
    
    /*
     * 	Here is double-checked locking pattern
     * 	-- To ensure there is only one thread that can access the resource in the synchronized block
     *  -- Predicate 1 cannot be removed, otherwise each time that a thread accesses here will grab the lock, which is unnecessary once the private field "instance" is assigned an instance
     *  -- Predicate 2 cannot be removed, otherwise 2 or more threads may pass the predicate 1 and wait to grab the lock to assign instances to the private field "instance" repeatedly 
     */
    if (instance == null || instance.equals(null)) {          // Predicate 1
      this.synchronized {
        if (instance == null || instance.equals(null)) {      // Predicate 2
          val clazz = ReflectionUtils.getJavaClassFromScalaType[T]
          instance = clazz.newInstance()
        }
      }
    }
    
    /*	Return the singleton instance	 */
    instance
  }
  
  /**
   * 	This is a method to get the singleton instance for the compatibility with Java
   */
  def getInstance(clazz: Class[_]) = {
    
    /*
     * 	Here is double-checked locking pattern
     * 	-- To ensure there is only one thread that can access the resource in the synchronized block
     *  -- Predicate 1 cannot be removed, otherwise each time that a thread accesses here will grab the lock, which is unnecessary once the private field "instance" is assigned an instance
     *  -- Predicate 2 cannot be removed, otherwise 2 or more threads may pass the predicate 1 and wait to grab the lock to assign instances to the private field "instance" repeatedly 
     */
    if (instance == null || instance.equals(null)) {          // Predicate 1
      this.synchronized {
        if (instance == null || instance.equals(null)) {      // Predicate 2
          instance = clazz.newInstance()
        }
      }
    }
    
    /*	Return the singleton instance	 */
    instance
  }
}

File: JavaAndScala/src/main/java/com/java/se/conclusion/design/factory/singleton/LazySingletonFactory.java
package com.java.se.conclusion.design.factory.singleton;

/**
 * 	This is a class to implement lazy singleton factory pattern
 * 	-- Declare a private, static and Object-typed reference
 *  -- Declare a public and static getter to get this unique object:
 *     -- Double-checked locking pattern is used in this pattern
 *     
 * @author VinceYuan
 *
 */
public class LazySingletonFactory {

	/**
	 * 	Declare a private, static and Object-typed reference
	 */
	private static volatile Object INSTANCE;
	
	/**
	 * 	Declare a public and static getter to get this unique object: this should be used on or after J2SE 5.0
	 *	-- If the object is not instantiated, then create an instance
	 *	-- If the object has been instantiated, then directly return the instance
	 * @param clazz
	 * @return
	 */
	public static Object getInstance(Class<?> clazz) {
		
		/*	If the object is not instantiated, then create an instance	 */
		if (INSTANCE == null || INSTANCE.equals(null)) {
			/**
			 * 	Here is double-checked locking pattern
			 *  -- Make sure there is always only one thread accessing the resource in this synchronized block
			 *  -- Once the instance is created, this synchronized block will never be accessed
			 */
			synchronized (LazySingletonFactory.class) {
				if (INSTANCE == null || INSTANCE.equals(null)) {
					try {
						INSTANCE = clazz.newInstance();
					} catch (InstantiationException | IllegalAccessException e) {
						e.printStackTrace();
					}
				}
			}
		}
		
		/*	Return the instance	 */
		return INSTANCE;
	}
}
